paddle/scripts/docker/README.md

```diff
@@ -1,38 +1,146 @@
+   ```
+
+## Goal
+
+We want two Docker images for each version of PaddlePaddle:
+
+1. `paddle:<version>-dev`
+
+   This a development image contains only the development tools.  This standardizes the building tools and procedure.  Users include:
+
+   - developers -- no longer need to install development tools on the host, and can build their current work on the host (development computer).
+   - release engineers -- use this to build the official release from certain branch/tag on Github.com.
+   - document writers / Website developers -- Our documents are in the source repo in the form of .md/.rst files and comments in source code.  We need tools to extract the information, typeset, and generate Web pages.
+
+   So the development image must contain not only source code building tools, but also documentation tools:
+
+   - gcc/clang
+   - nvcc
+   - Python
+   - sphinx
+   - woboq
+   - sshd
+
+   where `sshd` makes it easy for developers to have multiple terminals connecting into the container.
+
+1. `paddle:<version>`
+
+   This is the production image, generated using the development image. This image might have multiple variants:
+
+   - GPU/AVX   `paddle:<version>-gpu`
+   - GPU/no-AVX  `paddle:<version>-gpu-noavx`
+   - no-GPU/AVX  `paddle:<version>`
+   - no-GPU/no-AVX  `paddle:<version>-noavx`
+
+   We'd like to give users choices of GPU and no-GPU, because the GPU version image is much larger than then the no-GPU version.
+
+   We'd like to give users choices of AVX and no-AVX, because some cloud providers don't provide AVX-enabled VMs.
+
+## Dockerfile
+
+To realize above goals, we need only one Dockerfile for the development image.  We can put it in the root source directory.
+
+Let us go over our daily development procedure to show how developers can use this file.
+
+1. Check out the source code
 
-1. 用dev image build 我们的源码：
    ```bash
-   docker run -it -p 2022:22 -v $PWD:/paddle paddlepaddle/paddle:dev-0.10.0rc1  /paddle/build.sh
-   ```  
-   注意，这里的 `-v ` 参数把host上的源码目录里的内容映射到了container里的`/paddle` 目录；而container里的 `/paddle/build.sh` 就是源码目录里的 `build.sh`。上述命令调用了本地源码中的 bulid.sh 来build了本地源码，结果在container里的 `/paddle/build` 目录里，也就是本地的源码目录里的 `build` 子目录。
+   git clone https://github.com/PaddlePaddle/Paddle paddle
+   ```
+
+1. Do something
 
-1. 我们希望上述 `build.sh` 脚本在 `build` 子目录里生成一个Dockerfile，使得我们可以运行：
    ```bash
-   docker build -t paddle  ./build
+   cd paddle
+   git checkout -b my_work
+   Edit some files
    ```
-   来生成我们的production image。
-   
-1. 有了这个production image之后，我们可能会希望docker push 到dockerhub.com的我们自己的名下，然后可以用来启动本地或者远程（Kubernetes）jobs：
+
+1. Build/update the development image (if not yet)
 
    ```bash
-   docker tag paddle yiwang/paddle:did-some-change
-   docker push
-   paddlectl run yiwang/paddle:did-some-change /paddle/demo/mnist/train.py
+   docker build -t paddle:dev . # Suppose that the Dockerfile is in the root source directory.
+   ```
+
+1. Build the source code
```

Comment:
Should we build paddle when building the develop docker image? So developers will get to work immediately before they make some modifications and rebuild paddle again?